crates/bevy_ecs/src/schedule_v3/graph_utils.rs

```diff
@@ -0,0 +1,219 @@
+use std::fmt::Debug;
+
+use bevy_utils::{
+    petgraph::{graphmap::NodeTrait, prelude::*},
+    HashMap, HashSet,
+};
+use fixedbitset::FixedBitSet;
+
+use crate::schedule_v3::set::*;
+
+/// Unique identifier for a system or system set.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum NodeId {
+    System(usize),
+    Set(usize),
+}
+
+impl NodeId {
+    /// Returns the internal integer value.
+    pub fn index(&self) -> usize {
+        match self {
+            NodeId::System(index) | NodeId::Set(index) => *index,
+        }
+    }
+
+    /// Returns `true` if the identified node is a system.
+    pub const fn is_system(&self) -> bool {
+        matches!(self, NodeId::System(_))
+    }
+
+    /// Returns `true` if the identified node is a system set.
+    pub const fn is_set(&self) -> bool {
+        matches!(self, NodeId::Set(_))
+    }
+}
+
+/// Specifies what kind of edge should be inserted in the dependency graph.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub(crate) enum DependencyEdgeKind {
+    /// A node that should be preceded.
+    Before,
+    /// A node that should be succeeded.
+    After,
+}
+
+/// Configures ambiguity detection for a single system.
+#[derive(Clone, Debug, Default)]
+pub(crate) enum Ambiguity {
+    #[default]
+    Check,
+    /// Ignore warnings with systems in any of these system sets. May contain duplicates.
+    IgnoreWithSet(Vec<BoxedSystemSet>),
+    /// Ignore all warnings.
+    IgnoreAll,
+}
+
+#[derive(Clone)]
+pub(crate) struct GraphInfo {
+    pub(crate) sets: Vec<BoxedSystemSet>,
+    pub(crate) dependencies: Vec<(DependencyEdgeKind, BoxedSystemSet)>,
```

Comment:
The tuple struct here was hard to understand elsewhere: can we swap to a trivial named struct?